### ISSUE TITLE
New, bigger, laser engravers

### ISSUE THREAD

**Issue by GilDev** [camrein/EzGraver#43]
Hello,

This is just for discussion. I ran across this [1 W engraver](https://fr.aliexpress.com/item/1000mW-Automatic-DIY-Laser-Engraving-Machine-Carving-Engraver-Carver-AC100-240V-Miniature-DIY-Handicraft-Wood-Burning/32841140848.html) (there is also a [1.5 W black version](https://fr.aliexpress.com/item/1500-mW-USB-Laser-Graveur-DIY-Logo-Marque-Imprimante-Cutter-Carver-Gravure-Laser-Sculpture-Machine-pour/32847171490.html)) and it seems to be quite good, with a larger engraving area (80x80 mm) than NEJE's and you can put it above the objects you'd like to engrave.

I don't know the brand, nor the protocol and software used for these engravers, but if it is ever supported by EzGraver, that would be awesome!

Of course I know it's not the priority and it may be quite complicated to make EzGraver work with new engravers, but we never know what might happen in the future. If I get one I will seriously try to contribute to EzGraver's source to try to make it work.

**Comment by harlleysathler** [comment]
Interesting machines. Found them on GearBest, under a Decaker brand name.

I liked the buttons in a place more accessible than NEJE and the green acrylic panel on front. The double work area is a plus, obviously, but the "put it above the objects" thing can be achieved by modifying the NEJE machine. 

I saw a video on you tube in witch a french guy transport the X axis assembly to the bottom: [https://www.youtube.com/watch?v=MbpyWDcFjrs&t=5s](https://www.youtube.com/watch?v=MbpyWDcFjrs&t=5s)

**Comment by GilDev** [comment]
That's interesting, but of course it's better if it's made that way from the beginning.
And these engravers cost the same amount than my NEJE's at the time I bought it.

**Comment by harlleysathler** [comment]
Of course! I'm not a big fan of this type of modification. Considering the cost (in fact the 1500mW version costs a bit less than I paid for my NEJE), I ordered one to try!

**Comment by GilDev** [comment]
Cool! Please post your experiences when you've tried it! =)

**Comment by camrein** [comment]
Thanks for sharing the new devices. Of course, they look interesting. A friend of mine even applied the mod the french guy after this post. :)
Of course, extending the support for more devices would be great. Primarily, I do not think adding support for additional devices is difficult to accomplish. However, what it makes it difficult for me is "reverse engineering" devices which I do not have direct access to. Nevertheless, I'm absolutely overwhelmed by the support of the people.

I was playing with the idea of stronger separating the protocol implementations to make it easier to add support for new devices (maybe even embed a script langauge or something like that so trial and error for protocol implementations is straightforward). But before I want to finish the support for v3 and the strongly needed code cleanup coming with it.

**Comment by GilDev** [comment]
All of that seems awesome. =) EzGraver has a great life ahead, I hope I will be able to contribute at some point.

**Comment by lonelymyp** [comment]
I have such a laser. And the standard software is bad enough.
At first glance the protocol is very similar, I looked with a sniffer.
The difference in the control commands, they are slightly different (for example, the command "left" is not 245, but 134), 
and in that the picture is sent in parts (perhaps because the 1050x1050 is not entirely fit in memory).

I would really like to help add support to this laser

**Comment by lonelymyp** [comment]
Brief description of the control commands.

Write | description  | read |  description
-- | -- | -- | --
$40 | connect| B10 |  
  |   | F8 P1.000000 | Current power 1
  |   | F9 P0.203000 | Current time 0.203
  |   | connect ----- | 
$10 P0 | Fan Off |   |  
$10 P1000 | Fan On |   |  
$20 P563 398   51 71 | Start preview  | F10 |  
$25 P285   285 | Stop preview  | F11 |  
$25 P1401   244 | Go to:   X70.07 Y12.20 | F11 |  
$25 P1600   0 | Go to:   X80.00 Y00.00 | F11 |  
$25 P1600   1600 | Go to:   X80.00 Y80.00 | F11 |  
$9 P0.000 | Time 0% |   |  
$9 P1.000 | Time 100% |   |  
$9 P0.500 | Time 50% |   |  
$8 P0.000 | Power 0% |   |  
$8 P1.000 | Power 100% |   |

**Comment by GilDev** [comment]
That is great! Thank you for your contribution, I'm sure it will help @camrein a lot!
I think some code rework must be done to allow for new protocols to be easily implemented (maybe a specific structure in JSON for example, linking actions and control commands).
I'm more and more interested in this engraver, I may exchange my NEJE's DK-8-KZ for this one. =)

**Comment by lonelymyp** [comment]
Here is the process of engraving:
$30 - start command
P780 780 - start point XY
P2 - ???
F20 - start Ok
B1 - data packet request

![log](https://user-images.githubusercontent.com/2085051/35354265-91c7a82c-015a-11e8-978e-91790a17df48.png)


Here is an archive with a test picture and three part data packets:
[test.zip](https://github.com/camrein/EzGraver/files/1661473/test.zip)

**Comment by dk5ras** [comment]
Hi! 

I just got a KKMoon laser engraver, 3W, with some crappy software, and I wonder if implementing this model is already being worked on. Or can I do something to assist, maybe sniffing commands or so? Or giving on of you the original software to look into it?

Ralph.

**Comment by ekaitz-zarraga** [comment]
hi,
I'm [migrating EzGraver to python](https://github.com/ekaitz-zarraga/pygraver) and I'd like to add suport for Decaker and some other protocols. @lonelymyp can you give me more information of your findings?

**Comment by lonelymyp** [comment]
@ekaitz-zarraga Yes, of course, I am ready to help. Ask questions.
The control protocol is quite simple, we only need to figure out how the picture is encrypted for transmission. There are certain patterns in the transmitted string.

**Comment by ekaitz-zarraga** [comment]
Let's move the discussion to my project. Once it's done I'll probably come back with a patch.
